### ISSUE TITLE
FetchSize optimization possibility for Oracle

### ISSUE THREAD

**Issue by salixzs** [DapperLib/Dapper#654]
There are data retrieval optimization possibility for Oracle database Command and DataReader to specify DataFetch property for any of those (DataReader inherits from Command) to optimize (make faster) amount of data fetched from DataReader in one database trip. 
There are numerous examples on how it can be achieved and set when using pure ADO.NET, but in essence one needs to have to set this parameter for DataReader just before start to read from it.
Some examples are here:
https://kiquenet.wordpress.com/2016/06/18/performance-dataaccess-oracle-and-odp-net/
https://www.codeproject.com/tips/428152/oracledatareader-fetchsize-property
but you can find in Google more examples and explanations with keywords "ODP.NET FetchSize".

```c#
OracleConnection myConnection = new OracleConnection(myConnectionString);
OracleCommand myCommand = new OracleCommand(mySelectQuery, myConnection);
myConnection.Open();
using (OracleDataReader reader = myCommand.ExecuteReader(CommandBehavior.CloseConnection))
{
    // here goes the trick
    // lets get 1000 rows on each round trip
    reader.FetchSize = reader.RowSize * 1000;

    while (reader.Read())
    {
        // reads the records normally
    }
}// close and dispose stuff here
```

It would be nice to somehow set FetchSize to ```reader.RowSize * 1``` for QueryFirst* QuerySingle and ```reader.RowSize * 100``` for Query (returning multiple rows).

Other option would be to give possibility to specify it via some hooking.

Do you see it is possible at all?

**Comment by mgravell** [comment]
It isn't inconceivable; we already do some per-type hacks for some other
things. Can I ask what kind of performance difference were talking about
here in real world scenarios?

**Comment by salixzs** [comment]
Here are some links to blog articles people measuring differences in data retrieval speeds if FetchSize is adjusted:
http://metekarar.blogspot.com/2013/04/performance-improvement-for-odpnet.html
http://grow-n-shine.blogspot.com/2011/05/optimize-net-access-with-oracle.html
http://guyharrison.squarespace.com/blog/2009/10/19/oracle-performance-programming-net.html

Here is ORACLE Video from Oracle Virtual Technology Summit 2015
https://www.youtube.com/watch?v=ozMPGmsKcoA
Start around 10:14, demo on it around 11:08

Here is marginally related article of data retrieval over different networks and how FetchSize affects it:
http://markbairden.blogspot.com/2012/03/database-performance-measuring-effects.html

**Comment by vfabregat** [comment]
Hi! I having the same issue, I cannot configure it, I also tried with the xml configuration... As an idea, is possible to add to the C`ommandDefinition` like the property `InitialLONGFetchSize`.  [here is the code](https://github.com/StackExchange/Dapper/blob/a213b2aeee66ac0bff327ff19e70516fb629abe3/Dapper/CommandDefinition.cs#L143).

Or there any mechanism to change the `CommandDefinition` by a custom one to add the desired behaviour?

**Comment by vfabregat** [comment]
For future references, finally, I change the FetchSize using the xml configuration [here](https://docs.oracle.com/cd/E48297_01/doc/win.121/e41125/featConfig.htm#autoId4) for more info

**Comment by NickCraver** [comment]
We might add this in settings for v3, or perhaps we point people at Dapper.Oracle to house all of the oddities about their provider in one spot. For example, they already have support for this in their parameters: https://github.com/DIPSAS/Dapper.Oracle/blob/27809534024117ad493f2cb25f1a60d906f85ee7/src/Dapper.Oracle/OracleDynamicParameters.cs

**Comment by epaulsen** [comment]
Maintainer of "Oracle oddities" (a.k.a Dapper.Oracle) here 😄 
At current, we don't support modifying the datareader, because there is no extension point for it in Dapper.  What we do support, is setting the initial fetch size for binaries(InitialLOBFetchSize)

As the OP mentions, this needs to happen right after the reader is created, before any reads are performed.
This would really be a great improvement for performance tuning against Oracle, so I have created a PR to add a hook that will enable external modification of the IDataReader created inside Dapper.  
See pullrequest #1505 for details.

**Comment by WillReichl** [comment]
Hey guys. I could be wrong here, but I seem to get equivalent throughput from Dapper, as when using an OracleDataReader w/ fetch size configured, when I use the "OracleConfiguration" API to set fetch size.

Seems like you can set it before or after you create the connection.

Your mileage may vary.

![image](https://user-images.githubusercontent.com/43020530/123665366-370e0080-d806-11eb-8973-8f21885ebfe0.png)

https://github.com/oracle/dotnet-db-samples/blob/master/samples/dotnet-core/configuration-api/configuration-class.cs

**Comment by mgravell** [comment]
static configuration data isn't something a library should be defining - as it impacts all callers, regardless of whether they're going via our library; if consumers want to do that: great!

**Comment by epaulsen** [comment]
Agree with @mgravell on this.  Also, the normal use-case for setting fetchsize is directly on the reader, based on rowsize and an estimated number of rows.  Setting it via static initialization might get you halfway there, but it will never be optimal for all queries in a dataaccess layer.

**Comment by WillReichl** [comment]
Yep, agree as well! Sorry guys, I intended my comment to be consumer-facing, not library-facing, should anyone else have run into the same network performance roadblock that I did and wanted a potential workaround rather than falling back on using ODP.NET libraries.

Thanks for all you do.

**Comment by salixzs** [comment]
Do I understand this is also fixed by [2.0.151](https://github.com/DapperLib/Dapper/releases/tag/2.0.151)
